src/abstractdataframe/where.jl

```diff
@@ -0,0 +1,191 @@
+using DataFrames: normalize_selection, ColumnIndex
+# where allows a transformation specification without a target column name or a column
+
+_process_where_pair(i::Int, a::ColumnIndex) = a => Symbol(:x, i)
+_process_where_pair(i::Int, @nospecialize(a::Pair{<:Any, <:Base.Callable})) =
+    first(a) => last(a) => Symbol(:x, i)
+_process_where_pair(i::Int, a) =
+    throw(ArgumentError("Condition specifier $a is not supported by `where`"))
+
+function _get_where_conditions(df::Union{AbstractDataFrame, GroupedDataFrame},
+                               @nospecialize(args), dropmissing::Bool)
+    conditions = Any[_process_where_pair(i, a) for (i, a) in enumerate(args)]
+
+    isempty(conditions) && throw(ArgumentError("At least one condition must be passed"))
+
+    if df isa AbstractDataFrame
+        df_conditions = select(df, conditions..., copycols=!(df isa DataFrame))
+    else
+        df_conditions = select(df, conditions...,
+                               copycols=!(parent(df) isa DataFrame), keepkeys=false)
+    end
+    test_type = dropmissing ? Union{Missing, Bool} : Bool
+    for col in eachcol(df_conditions)
+        if !(eltype(col) <: test_type)
+            throw(ArgumentError("each transformation must return a vector whose " *
+                                "eltype is subtype of $test_type"))
+        end
+    end
+    @assert ncol(df_conditions) == length(conditions)
+    if ncol(df) == 1
+        return isequal.(df_conditions[!, 1], true)
+    else
+        return isequal.(.&(eachcol(df_conditions)...), true)
+    end
```

Comment:
It would probably be better to use `===` instead of `isequal`, as core devs strongly discourage the introduction of "thruthiness" in Julia (that is, objects which are treated as `true` but aren't `Bool`). It shouldn't make a difference in existing cases but it prevents non-standard uses that would have a different behavior than `if`.

Reply:
OK - changed to `===`.